### ISSUE TITLE
Handle empty error responses from OAuth APIs

### ISSUE THREAD

**Issue by dmayo3** [lepture/flask-oauthlib#357]
We're using this library with a 3rd party OAuth API.

Sometimes we get an error response back from their API that appears to have an empty response body but supplies `Content-Type: application/json`. This leads to a parsing error.

Clearly the API is behaving badly and I've filed an issue with them, but it should be possible for Flask OAuthlib to tolerate this scenario or enable the developer to suppress automatic parsing on a per request-basis.

It looks like it's possible to configure flask-oauthlib to override the Content-Type to text/plain and thus probably workaround the issue. However, this is done an construction time and so affects all requests in the application that use the same object.

Here is the relevant portion of the trace we get:

```
File "/usr/local/lib/python3.6/site-packages/flask_oauthlib/client.py", line 390, in get
  return self.request(*args, **kwargs)
File "/usr/local/lib/python3.6/site-packages/flask_oauthlib/client.py", line 478, in request
  return OAuthResponse(resp, content, self.content_type)
File "/usr/local/lib/python3.6/site-packages/flask_oauthlib/client.py", line 166, in __init__
  content_type=content_type,
File "/usr/local/lib/python3.6/site-packages/flask_oauthlib/client.py", line 122, in parse_response
  return json.loads(content)
File "/usr/local/lib/python3.6/site-packages/flask/json.py", line 146, in loads
  return _json.loads(s, **kwargs)
File "/usr/local/lib/python3.6/site-packages/simplejson/__init__.py", line 533, in loads
  return cls(encoding=encoding, **kw).decode(s)
File "/usr/local/lib/python3.6/site-packages/simplejson/decoder.py", line 370, in decode
  obj, end = self.raw_decode(s)
File "/usr/local/lib/python3.6/site-packages/simplejson/decoder.py", line 400, in raw_decode
  return self.scan_once(s, idx=_w(s, idx).end())
simplejson.scanner.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by lepture** [comment]
@dmayo3 Are you using this library for client only?

**Comment by dmayo3** [comment]
@lepture yes, we're not using the OAuth Provider features since the 3rd party OAuth API implements that for us

**Comment by lepture** [comment]
@dmayo3 I'm working https://github.com/lepture/authlib to replace Flask-OAuthlib.

The client part is stable enough (more stable than Flask-OAuthlib) for production now. I'll release the first version soon. Stay tuned.

**Comment by dmayo3** [comment]
@lepture thanks

FWIW, I've realised some of the details I originally posted were wrong.

> Sometimes we get an error response back from their API that appears to have an empty response body but supplies Content-Type: application/json

Actually they weren't sending a `Content-Type` header at all, and the response contained some plain text.

The issue was being triggered by using `application/json` as the default type for parsing:

https://github.com/lepture/flask-oauthlib/blob/master/flask_oauthlib/client.py#L122

In this instance it might be safer to assume text/plain as the default, but then that might break other apps.

**Comment by lepture** [comment]
@dmayo3 Is it OAuth 1? The implementation is not good enough. The problem should be here:

https://github.com/lepture/flask-oauthlib/blob/master/flask_oauthlib/client.py#L595
https://github.com/lepture/flask-oauthlib/blob/master/flask_oauthlib/client.py#L644

**Comment by lepture** [comment]
@dmayo3 Authlib 0.2 is working stable enough for client part now. Take a look at the documentation:
https://docs.authlib.org/en/v0.2/client/frameworks.html#module-authlib.client.flask

**Comment by dmayo3** [comment]
@lepture OK, thanks

In response to your question we're using OAuth 2.0, but in this case we were just making a simple GET call to the 3rd party API to obtain some additional information about the user.

This was the call sequence: `get() -> request() -> OAuthResponse() -> parse_response()`

**Comment by lepture** [comment]
@dmayo3 https://github.com/lepture/flask-oauthlib/blob/master/flask_oauthlib/client.py#L506

You can set a `content_type` to overwrite the content type.

**Comment by dmayo3** [comment]
Yes, but not for just that request.

> It looks like it's possible to configure flask-oauthlib to override the Content-Type to text/plain and thus probably workaround the issue. However, this is done an construction time and so affects all requests in the application that use the same object.

**Comment by lepture** [comment]
@dmayo3 A workaround for now could be:

1. set content_type to plain/text before requesting
2. set content_type back after requesting
